File: src/main/java/org/sonar/plugins/xquery/checks/StrongTypingInFLWORCheck.java
/*
 * © 2013 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks;

import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.plugins.xquery.parser.XQueryParser;
import org.sonar.plugins.xquery.parser.XQueryTree;

/**
 * Checks for strong typing in module variable declarations
 * 
 * @since 1.0
 */
@Rule(
    key = "StrongTypingInFLWOR",
    name = "Use Strong Typing FLWOR Expressions",
    description = "Declare types for FLWOR 'let' and 'for' clauses to increase readability and catch potential bugs.  Also try to scope the types as narrowly as possible (i.e. use 'element()' instead of 'item()' when the value is an element) and include quantifiers on each type. For more information see http://labslcl0274.qalab.ldschurch.org:8403/viewArticle.xqy?id=424",
    priority = Priority.MINOR
)
public class StrongTypingInFLWORCheck extends AbstractCheck {

    @Override
    public void enterExpression(XQueryTree node) {
        super.enterExpression(node);
        
        // Process FLWOR expressions
        if (XQueryParser.FLOWRExpr == node.getType()) {

            // If we have children we can loop through them and check both the
            // 'let' and 'for' clauses
            if (node.getChildCount() > 0) {
                for (XQueryTree clause : node.getChildren()) {                    

                    // Check any 'for' clauses
                    if ("ForClause".equals(clause.getText())) {
                        if (clause.find("ForType") == null) {
                            createViolation(clause.getLine());
                        }
                    }

                    // Check any 'let' clauses
                    if ("LetClause".equals(clause.getText())) {
                        if (clause.find("LetType") == null) {
                            createViolation(clause.getLine());
                        }
                    }
                }
            }
        }
    }
}

File: src/main/java/org/sonar/plugins/xquery/checks/StrongTypingInFLWORCheck.kt
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks

import org.antlr.v4.runtime.ParserRuleContext
import org.sonar.api.rule.RuleKey
import org.sonar.check.Rule
import org.sonar.plugins.xquery.checks.StrongTypingInFLWORCheck.Companion.RULE_KEY
import org.sonar.plugins.xquery.parser.XQueryParser.ForVarContext
import org.sonar.plugins.xquery.parser.XQueryParser.LetVarContext
import org.sonar.plugins.xquery.parser.getLine
import org.sonar.plugins.xquery.rules.CheckClasses

/**
 * Checks for strong typing in module variable declarations
 *
 * @since 1.0
 */
@Rule(key = RULE_KEY)
class StrongTypingInFLWORCheck : AbstractCheck() {

    override fun enterExpression(node: ParserRuleContext) {
        super.enterExpression(node)

        when (node) {
            is ForVarContext -> if (node.type == null) createIssue(RULE, node.getLine())
            is LetVarContext -> if (node.type == null) createIssue(RULE, node.getLine())
        }
    }

    companion object {
        const val RULE_KEY = "StrongTypingInFLWOR"
        private val RULE = RuleKey.of(CheckClasses.REPOSITORY_KEY, RULE_KEY)
    }
}
